### PR TITLE
Replace string constants with a proper enum in FilterParser

### LINKED ISSUE #1077: Replace string constants with a proper enum in FilterParser
#1076 adds three string constants to `FilterParser`: `EXPECTED_ATTRIBUTE_NAME`, `EXPECTED_OPERATORS`, and `EXPECTED_VALUE`. These are used to indicate what made the parser fail. We *had to* use strings because Nom 5 doesn't allow anything else for its `context` combinator.

Nom 6 relaxes the requirement (https://github.com/Geal/nom/issues/1131), so once that's released, we should migrate and replace our three strings with a proper `enum`.

### PR DESCRIPTION
**Pull request by davbrito**
fixes #1077
I made a new error type for the parser and a combinator called `expect` which resembles the `context` combinator.

### CODE REVIEW

**Comment by Minoru** [comment]
Hey, thanks for working on this! I haven't looked at the code, but I see the commit history — please rebase your commits onto `master`. The merges you have right now make the history really ugly and hard to follow, I'd prefer not to have them. Thanks! (It looks like you've been using Git for a while, so you probably know how to fix this, but if any problems arise — please comment here, I'll try to help.)

**Comment by davbrito** [comment]
I'll rebase it when I'm done.

**Comment by Minoru** [comment]
Ah, okay!

What's your plan regarding #1260? You sure the upgrade to Nom 6 won't interfere with the work you're doing here? And vice versa — if someone fixes #1260, would that create merge conflicts for you here?

**Comment by davbrito** [comment]
> Ah, okay!
> 
> What's your plan regarding #1260? You sure the upgrade to Nom 6 won't interfere with the work you're doing here? And vice versa — if someone fixes #1260, would that create merge conflicts for you here?

I would like to fix #1269 
How do I do that?

**Comment by Minoru** [comment]
> I would like to fix #1260 How do I do that?

I see you've submitted #1268 against the #1260. That's clever :) And that'll work fine, so :+1:

**Comment by davbrito** [comment]
@Minoru Waiting for #1260 and #1268 to continue this

**Comment by Minoru** [comment]
I merged #1268, and Dependabot will merge #1260 automatically once all the CI checks are done (in ≈40 minutes).

**Comment by davbrito** [comment]
I think this is ready. Let's wait for the CI checks.

**Review by Minoru — CHANGES_REQUESTED** [review]
Looks good! I only have a couple general comments, please see below.

**Review by Minoru — APPROVED** [review]
Cool, thank you!

The change to rust/libnewsboat/Cargo.toml should've changed Cargo.lock as well. It's not worth another code review round-trip, so I'll just fix that myself before merging.

**Comment by Minoru** [comment]
A couple requests for the future:

1. please don't mark review comments resolved. I think it's fair that the person who raised the concern gets to decide if it was addressed or not; in other words, it's the reviewer who marks (their) comments resolved. If you need a marker for your own use, consider adding :eyes: reaction to the comments that you've processed already.

2. please don't force-push after a review got posted. GitHub's support for force-pushes is pretty useless: they add diffs between "old" and "new", but if you force-push a couple times in a row, only the last diff is kept. Once a review was published, it's better to add fixup commits (`git commit --fixup=ID`) which you can later squash automatically with `git rebase --autosquash --interactive …`.

    It didn't hurt for this PR, but for large ones with a handful of commits and the changes spread over a dozen files, it becomes *extremely* tiring to re-review everything after a force-push.

    If I see a fixup commit in the history, I'll give you time to squash them *after* the review is done, *before* I merge the PR.

By the way, I submit changes via PRs too, and mine are always open for everyone to review (for three days usually). So if you're interested to try out these rules from the reviewer's side, you're always welcome to drop a comment ;)

**Comment by davbrito** [comment]
> A couple requests for the future:
> 
> 1. please don't mark review comments resolved. I think it's fair that the person who raised the concern gets to decide if it was addressed or not; in other words, it's the reviewer who marks (their) comments resolved. If you need a marker for your own use, consider adding eyes reaction to the comments that you've processed already.
> 2. please don't force-push after a review got posted. GitHub's support for force-pushes is pretty useless: they add diffs between "old" and "new", but if you force-push a couple times in a row, only the last diff is kept. Once a review was published, it's better to add fixup commits (`git commit --fixup=ID`) which you can later squash automatically with `git rebase --autosquash --interactive …`.
>    It didn't hurt for this PR, but for large ones with a handful of commits and the changes spread over a dozen files, it becomes _extremely_ tiring to re-review everything after a force-push.
>    If I see a fixup commit in the history, I'll give you time to squash them _after_ the review is done, _before_ I merge the PR.
> 
> By the way, I submit changes via PRs too, and mine are always open for everyone to review (for three days usually). So if you're interested to try out these rules from the reviewer's side, you're always welcome to drop a comment ;)

Hi @minoru,
Thank you for your comments and recommendations, I will take everything into account in the future. I am somewhat new to this and every tip is very helpful. Thanks for everything.
